File: scripts/oncall-triage-groq.ts
#!/usr/bin/env bun

declare global {
  var process: {
    env: Record<string, string | undefined>;
  };
}

interface GitHubIssue {
  number: number;
  title: string;
  body: string | null;
  labels: Array<{ name: string }>;
  created_at: string;
  updated_at: string;
  comments: number;
  user: { id: number; login: string };
  pull_request?: {
    url: string;
  };
  reactions: {
    total_count: number;
    '+1': number;
    '-1': number;
    laugh: number;
    hooray: number;
    confused: number;
    heart: number;
    rocket: number;
    eyes: number;
  };
}

interface GitHubComment {
  id: number;
  body: string;
  created_at: string;
  reactions: {
    total_count: number;
  };
}

interface GroqMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

interface IssueAnalysis {
  issueNumber: number;
  title: string;
  shouldLabel: boolean;
  reason: string;
  isBorderline?: boolean;
}

async function githubRequest<T>(
  endpoint: string,
  token: string,
  method: string = 'GET',
  body?: any
): Promise<T> {
  const response = await fetch(`https://api.github.com${endpoint}`, {
    method,
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: "application/vnd.github.v3+json",
      "User-Agent": "oncall-triage-groq-script",
      ...(body && { "Content-Type": "application/json" }),
    },
    ...(body && { body: JSON.stringify(body) }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `GitHub API request failed: ${response.status} ${response.statusText} - ${errorText}`
    );
  }

  return response.json();
}

async function groqRequest(
  messages: GroqMessage[],
  apiKey: string,
  model: string = 'llama-3.3-70b-versatile'
): Promise<string> {
  const response = await fetch('https://api.groq.com/openai/v1/chat/completions', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      model,
      messages,
      temperature: 0.3,
      max_tokens: 2000,
    }),
  });

  if (!response.ok) {
    const errorText = await response.text();
    throw new Error(
      `Groq API request failed: ${response.status} ${response.statusText} - ${errorText}`
    );
  }

  const data = await response.json();
  return data.choices[0].message.content;
}

async function getRecentlyUpdatedIssues(
  owner: string,
  repo: string,
  token: string,
  daysBack: number = 3
): Promise<GitHubIssue[]> {
  const cutoffDate = new Date();
  cutoffDate.setDate(cutoffDate.getDate() - daysBack);
  
  const issues: GitHubIssue[] = [];
  let page = 1;
  const perPage = 100;
  
  while (true) {
    const pageIssues = await githubRequest<GitHubIssue[]>(
      `/repos/${owner}/${repo}/issues?state=open&sort=updated&direction=desc&per_page=${perPage}&page=${page}`,
      token
    );
    
    if (pageIssues.length === 0) break;
    
    let foundOldIssue = false;
    for (const issue of pageIssues) {
      // Skip pull requests
      if (issue.pull_request) continue;
      
      const updatedAt = new Date(issue.updated_at);
      if (updatedAt < cutoffDate) {
        foundOldIssue = true;
        break;
      }
      
      issues.push(issue);
    }
    
    if (foundOldIssue || pageIssues.length < perPage) break;
    page++;
  }
  
  console.log(`Found ${issues.length} issues updated in the last ${daysBack} days`);
  return issues;
}

async function getIssueComments(
  owner: string,
  repo: string,
  issueNumber: number,
  token: string
): Promise<GitHubComment[]> {
  const comments = await githubRequest<GitHubComment[]>(
    `/repos/${owner}/${repo}/issues/${issueNumber}/comments`,
    token
  );
  return comments;
}

async function analyzeIssueWithGroq(
  issue: GitHubIssue,
  comments: GitHubComment[],
  groqApiKey: string
): Promise<{ shouldLabel: boolean; reason: string; isBorderline: boolean }> {
  // Calculate total engagements
  const commentCount = comments.length;
  const reactionCount = issue.reactions.total_count + 
    comments.reduce((sum, c) => sum + c.reactions.total_count, 0);
  const totalEngagements = commentCount + reactionCount;
  
  // Check if it has bug label
  const hasBugLabel = issue.labels.some(label => 
    label.name.toLowerCase().includes('bug')
  );
  
  // Check if it already has oncall label
  const hasOncallLabel = issue.labels.some(label => 
    label.name.toLowerCase() === 'oncall'
  );
  
  if (hasOncallLabel) {
    return {
      shouldLabel: false,
      reason: 'Already has oncall label',
      isBorderline: false
    };
  }
  
  // Quick filters
  if (!hasBugLabel && !issue.title.toLowerCase().includes('bug') && 
      !issue.body?.toLowerCase().includes('bug')) {
    return {
      shouldLabel: false,
      reason: 'Not identified as a bug',
      isBorderline: false
    };
  }
  
  if (totalEngagements < 50) {
    return {
      shouldLabel: false,
      reason: `Insufficient engagement (${totalEngagements} < 50)`,
      isBorderline: totalEngagements >= 40
    };
  }
  
  // Use Groq to analyze if it's truly blocking
  const issueContent = `
Title: ${issue.title}

Body:
${issue.body || '(no description)'}

Comments (${comments.length}):
${comments.slice(0, 10).map((c, i) => `Comment ${i + 1}: ${c.body.substring(0, 500)}`).join('\n\n')}
${comments.length > 10 ? '\n... (additional comments omitted)' : ''}
  `.trim();
  
  const messages: GroqMessage[] = [
    {
      role: 'system',
      content: `You are an expert at triaging GitHub issues to identify critical blocking bugs that require immediate oncall attention.

Analyze the issue and determine if it meets ALL of these criteria:
1. It describes a bug (not a feature request or question)
2. It has sufficient engagement (>= 50 total comments + reactions) [already verified]
3. It is truly BLOCKING - prevents core functionality from working with no reasonable workaround

For criterion 3, look for severity indicators like:
- "crash", "stuck", "frozen", "hang", "unresponsive"
- "cannot use", "blocked", "broken", "unusable"
- Users explicitly stating they are blocked or cannot work

Be CONSERVATIVE. Only flag issues that truly prevent users from getting work done.
Issues with workarounds, minor inconveniences, or cosmetic problems should NOT be labeled.

Respond with ONLY a JSON object in this exact format:
{
  "is_blocking": true/false,
  "reason": "brief explanation of your decision",
  "confidence": "high/medium/low"
}`
    },
    {
      role: 'user',
      content: issueContent
    }
  ];
  
  const response = await groqRequest(messages, groqApiKey);
  
  // Parse the JSON response
  try {
    const jsonMatch = response.match(/\{[\s\S]*\}/);
    if (!jsonMatch) {
      throw new Error('No JSON found in response');
    }
    
    const analysis = JSON.parse(jsonMatch[0]);
    
    return {
      shouldLabel: analysis.is_blocking === true && analysis.confidence !== 'low',
      reason: analysis.reason,
      isBorderline: analysis.is_blocking === true && analysis.confidence === 'low'
    };
  } catch (error) {
    console.error(`Failed to parse Groq response for issue #${issue.number}:`, error);
    console.error('Response was:', response);
    return {
      shouldLabel: false,
      reason: 'Failed to analyze with AI',
      isBorderline: false
    };
  }
}

async function addOncallLabel(
  owner: string,
  repo: string,
  issueNumber: number,
  token: string
): Promise<void> {
  await githubRequest(
    `/repos/${owner}/${repo}/issues/${issueNumber}/labels`,
    token,
    'POST',
    { labels: ['oncall'] }
  );
  console.log(`✓ Added oncall label to issue #${issueNumber}`);
}

async function main() {
  const githubToken = process.env.GITHUB_TOKEN;
  const groqApiKey = process.env.GROQ_API_KEY;
  const repository = process.env.GITHUB_REPOSITORY;
  
  if (!githubToken) {
    throw new Error('GITHUB_TOKEN environment variable is required');
  }
  
  if (!groqApiKey) {
    throw new Error('GROQ_API_KEY environment variable is required');
  }
  
  if (!repository) {
    throw new Error('GITHUB_REPOSITORY environment variable is required');
  }
  
  const [owner, repo] = repository.split('/');
  
  console.log(`Starting oncall triage for ${owner}/${repo}...`);
  console.log('');
  
  // Get recently updated issues
  const issues = await getRecentlyUpdatedIssues(owner, repo, githubToken, 3);
  
  const analyses: IssueAnalysis[] = [];
  let labeled = 0;
  
  // Process each issue
  for (const issue of issues) {
    console.log(`\nEvaluating issue #${issue.number}: ${issue.title}`);
    
    try {
      // Get comments
      const comments = await getIssueComments(owner, repo, issue.number, githubToken);
      
      // Analyze with Groq
      const analysis = await analyzeIssueWithGroq(issue, comments, groqApiKey);
      
      analyses.push({
        issueNumber: issue.number,
        title: issue.title,
        shouldLabel: analysis.shouldLabel,
        reason: analysis.reason,
        isBorderline: analysis.isBorderline
      });
      
      if (analysis.shouldLabel) {
        await addOncallLabel(owner, repo, issue.number, githubToken);
        labeled++;
      } else {
        console.log(`  → Skip: ${analysis.reason}`);
      }
      
      // Rate limiting: wait a bit between issues
      await new Promise(resolve => setTimeout(resolve, 1000));
    } catch (error) {
      console.error(`  ✗ Error processing issue #${issue.number}:`, error);
      analyses.push({
        issueNumber: issue.number,
        title: issue.title,
        shouldLabel: false,
        reason: `Error: ${error instanceof Error ? error.message : String(error)}`
      });
    }
  }
  
  // Print summary
  console.log('\n' + '='.repeat(80));
  console.log('ONCALL TRIAGE SUMMARY');
  console.log('='.repeat(80));
  console.log(`\nTotal issues evaluated: ${analyses.length}`);
  console.log(`Issues labeled with 'oncall': ${labeled}`);
  
  if (labeled > 0) {
    console.log('\n📍 Issues that received the oncall label:');
    analyses
      .filter(a => a.shouldLabel)
      .forEach(a => {
        console.log(`  - #${a.issueNumber}: ${a.title}`);
        console.log(`    Reason: ${a.reason}`);
      });
  }
  
  const borderline = analyses.filter(a => a.isBorderline && !a.shouldLabel);
  if (borderline.length > 0) {
    console.log('\n⚠️  Close calls (borderline cases that did not qualify):');
    borderline.forEach(a => {
      console.log(`  - #${a.issueNumber}: ${a.title}`);
      console.log(`    Reason: ${a.reason}`);
    });
  }
  
  if (labeled === 0) {
    console.log('\n✓ No critical blocking issues requiring oncall attention were found.');
  }
  
  console.log('');
}

// Run the script
main().catch(error => {
  console.error('Fatal error:', error);
  process.exit(1);
});

File: .github/scripts/oncall-triage-groq.py
#!/usr/bin/env python3
"""
Oncall Issue Triage using GroqCloud API
This script identifies critical issues that need oncall attention using Groq's LLM API.
"""

import os
import sys
import json
import time
import re
from datetime import datetime, timedelta
from typing import List, Dict, Any
import requests

# Configuration
GROQ_API_KEY = os.environ.get("GROQ_API_KEY")
GITHUB_TOKEN = os.environ.get("GITHUB_TOKEN")
GITHUB_REPOSITORY = os.environ.get("GITHUB_REPOSITORY")
GROQ_API_URL = "https://api.groq.com/openai/v1/chat/completions"
GROQ_MODEL = "llama-3.3-70b-versatile"  # Using Llama 3.3 70B for better reasoning

if not GROQ_API_KEY:
    print("Error: GROQ_API_KEY environment variable is required")
    sys.exit(1)

if not GITHUB_TOKEN:
    print("Error: GITHUB_TOKEN environment variable is required")
    sys.exit(1)

if not GITHUB_REPOSITORY:
    print("Error: GITHUB_REPOSITORY environment variable is required")
    sys.exit(1)

# Parse repository owner and name
try:
    owner, repo = GITHUB_REPOSITORY.split("/")
except ValueError:
    print(f"Error: GITHUB_REPOSITORY must be in 'owner/repo' format, got: {GITHUB_REPOSITORY}")
    sys.exit(1)


def github_api_request(endpoint: str, method: str = "GET", data: Dict = None) -> Any:
    """Make a request to GitHub API"""
    url = f"https://api.github.com/{endpoint}"
    headers = {
        "Authorization": f"Bearer {GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3+json",
        "X-GitHub-Api-Version": "2022-11-28"
    }
    
    try:
        if method == "GET":
            response = requests.get(url, headers=headers)
        elif method == "POST":
            response = requests.post(url, headers=headers, json=data)
        elif method == "PATCH":
            response = requests.patch(url, headers=headers, json=data)
        else:
            raise ValueError(f"Unsupported method: {method}")
        
        response.raise_for_status()
        return response.json()
    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 403:
            print(f"GitHub API rate limit exceeded or forbidden: {e}")
        elif e.response.status_code == 404:
            print(f"GitHub API resource not found: {e}")
        else:
            print(f"GitHub API error: {e}")
        raise


def get_recent_issues(days: int = 3, max_pages: int = 10) -> List[Dict]:
    """Fetch issues updated in the last N days"""
    cutoff_date = datetime.utcnow() - timedelta(days=days)
    issues = []
    page = 1
    
    print(f"Fetching open issues updated in the last {days} days...")
    
    while page <= max_pages:
        endpoint = f"repos/{owner}/{repo}/issues"
        params = f"?state=open&sort=updated&direction=desc&per_page=100&page={page}"
        
        page_issues = github_api_request(endpoint + params)
        
        if not page_issues:
            break
        
        for issue in page_issues:
            # Skip pull requests
            if "pull_request" in issue:
                continue
                
            updated_at = datetime.strptime(issue["updated_at"], "%Y-%m-%dT%H:%M:%SZ")
            
            if updated_at < cutoff_date:
                print(f"Reached issues older than {days} days, stopping pagination")
                return issues
            
            issues.append(issue)
        
        page += 1
    
    print(f"Found {len(issues)} candidate issues")
    return issues


def get_issue_comments(issue_number: int) -> List[Dict]:
    """Fetch all comments for an issue"""
    endpoint = f"repos/{owner}/{repo}/issues/{issue_number}/comments"
    return github_api_request(endpoint)


def get_user_login(comment: Dict) -> str:
    """Safely extract user login from comment, handling deleted users"""
    user = comment.get('user')
    if not user:
        return 'deleted'
    return user.get('login', 'unknown')


def count_reactions(reactions: Dict) -> int:
    """Count total reactions from a reactions dict"""
    if not reactions:
        return 0
    return sum([
        reactions.get("+1", 0),
        reactions.get("-1", 0),
        reactions.get("laugh", 0),
        reactions.get("hooray", 0),
        reactions.get("confused", 0),
        reactions.get("heart", 0),
        reactions.get("rocket", 0),
        reactions.get("eyes", 0)
    ])


def count_engagements(issue: Dict, comments: List[Dict]) -> int:
    """Count total engagements (comments + reactions)"""
    total = len(comments)
    
    # Count reactions on the issue
    total += count_reactions(issue.get("reactions"))
    
    # Count reactions on comments
    for comment in comments:
        total += count_reactions(comment.get("reactions"))
    
    return total


def call_groq_api(messages: List[Dict[str, str]]) -> str:
    """Call Groq API for LLM reasoning"""
    headers = {
        "Authorization": f"Bearer {GROQ_API_KEY}",
        "Content-Type": "application/json"
    }
    
    payload = {
        "model": GROQ_MODEL,
        "messages": messages,
        "temperature": 0.3,
        "max_tokens": 2000,
    }
    
    try:
        response = requests.post(GROQ_API_URL, headers=headers, json=payload)
        response.raise_for_status()
        result = response.json()
        return result["choices"][0]["message"]["content"]
    except requests.exceptions.HTTPError as e:
        if e.response.status_code == 429:
            print(f"Groq API rate limit exceeded: {e}")
        elif e.response.status_code == 401:
            print(f"Groq API authentication failed: {e}")
        else:
            print(f"Groq API error: {e}")
        raise
    except KeyError as e:
        print(f"Unexpected Groq API response format: {e}")
        raise


def analyze_issue_with_groq(issue: Dict, comments: List[Dict]) -> Dict[str, Any]:
    """Use Groq LLM to analyze if an issue needs oncall attention"""
    
    # Prepare issue content
    issue_text = f"""Issue #{issue['number']}: {issue['title']}

Body:
{issue['body'] or 'No description provided'}

Labels: {', '.join([label['name'] for label in issue['labels']])}

Comments ({len(comments)} total):
"""
    
    # Add recent comments (last 5)
    for comment in comments[-5:]:
        user_login = get_user_login(comment)
        body = comment.get('body', '')
        if body:
            issue_text += f"\n- {user_login}: {body[:200]}..."
    
    # Create prompt for Groq
    prompt = f"""You are an oncall triage assistant. Analyze this GitHub issue and determine if it needs immediate oncall attention.

{issue_text}

An issue qualifies for oncall label if it meets ALL these criteria:
1. It's a bug (has "bug" label or describes bug behavior)
2. It has at least 50 engagements (comments + reactions)
3. It's truly blocking (prevents core functionality, users can't work around it)

Blocking severity indicators: "crash", "stuck", "frozen", "hang", "unresponsive", "cannot use", "blocked", "broken"

Be conservative - only flag truly critical issues that prevent users from working.

Respond ONLY with a JSON object in this exact format:
{{
  "needs_oncall": true/false,
  "reason": "brief explanation of your decision",
  "is_bug": true/false,
  "is_blocking": true/false
}}"""

    messages = [
        {"role": "system", "content": "You are a technical triage assistant. Respond only with valid JSON."},
        {"role": "user", "content": prompt}
    ]
    
    response = call_groq_api(messages)
    
    # Parse JSON response
    try:
        # Try to extract JSON from the response
        response = response.strip()
        if "```json" in response:
            response = response.split("```json")[1].split("```")[0].strip()
        elif "```" in response:
            response = response.split("```")[1].split("```")[0].strip()
        
        result = json.loads(response)
        return result
    except json.JSONDecodeError as e:
        print(f"Failed to parse Groq response: {response}")
        print(f"Error: {e}")
        return {"needs_oncall": False, "reason": "Failed to parse response", "is_bug": False, "is_blocking": False}


def add_oncall_label(issue_number: int) -> bool:
    """Add oncall label to an issue"""
    try:
        endpoint = f"repos/{owner}/{repo}/issues/{issue_number}/labels"
        github_api_request(endpoint, method="POST", data={"labels": ["oncall"]})
        return True
    except Exception as e:
        print(f"Failed to add oncall label to issue #{issue_number}: {e}")
        return False


def main():
    print(f"Starting oncall triage for {GITHUB_REPOSITORY}")
    print(f"Using Groq model: {GROQ_MODEL}")
    print("=" * 80)
    
    # Fetch recent issues
    issues = get_recent_issues(days=3)
    
    if not issues:
        print("No issues to process")
        return
    
    processed_count = 0
    labeled_count = 0
    labeled_issues = []
    close_calls = []
    
    for issue in issues:
        issue_number = issue["number"]
        print(f"\nProcessing issue #{issue_number}: {issue['title']}")
        
        # Check if already has oncall label
        labels = [label["name"] for label in issue["labels"]]
        if "oncall" in labels:
            print(f"  Already has oncall label, skipping")
            continue
        
        # Check if it's a bug - check exact label match and case-insensitive title/body
        has_bug_label = any(label.lower() == "bug" for label in labels)
        title_lower = issue["title"].lower()
        body_lower = (issue["body"] or "").lower()
        
        # Look for bug-related terms (as whole words)
        # Including common indicators: bug, defect, error, crash, failure
        bug_pattern = r'\b(bug|defect|error|crash|failure)\b'
        has_bug_mention = has_bug_label or bool(re.search(bug_pattern, title_lower)) or bool(re.search(bug_pattern, body_lower))
        
        if not has_bug_mention:
            print(f"  Not a bug, skipping")
            continue
        
        # Get comments
        comments = get_issue_comments(issue_number)
        
        # Count engagements
        engagements = count_engagements(issue, comments)
        print(f"  Engagements: {engagements}")
        
        if engagements < 50:
            print(f"  Not enough engagements (need 50+), skipping")
            continue
        
        # Analyze with Groq
        print(f"  Analyzing with Groq...")
        analysis = analyze_issue_with_groq(issue, comments)
        
        processed_count += 1
        
        print(f"  Analysis: {analysis['reason']}")
        print(f"  Is bug: {analysis['is_bug']}, Is blocking: {analysis['is_blocking']}, Needs oncall: {analysis['needs_oncall']}")
        
        if analysis["needs_oncall"]:
            print(f"  ✓ Adding oncall label")
            if add_oncall_label(issue_number):
                labeled_count += 1
                labeled_issues.append({
                    "number": issue_number,
                    "title": issue["title"],
                    "reason": analysis["reason"]
                })
        else:
            if engagements >= 50 and analysis["is_bug"]:
                close_calls.append({
                    "number": issue_number,
                    "title": issue["title"],
                    "reason": analysis["reason"]
                })
        
        # Rate limiting
        time.sleep(1)
    
    # Print summary
    print("\n" + "=" * 80)
    print("ONCALL TRIAGE SUMMARY")
    print("=" * 80)
    print(f"Total candidate issues evaluated: {processed_count}")
    print(f"Issues that received oncall label: {labeled_count}")
    
    if labeled_issues:
        print("\nIssues labeled with 'oncall':")
        for item in labeled_issues:
            print(f"  - #{item['number']}: {item['title']}")
            print(f"    Reason: {item['reason']}")
    
    if close_calls:
        print("\nClose calls (didn't quite meet criteria):")
        for item in close_calls:
            print(f"  - #{item['number']}: {item['title']}")
            print(f"    Reason: {item['reason']}")
    
    if not labeled_issues and not close_calls:
        print("\nNo issues qualified for oncall label.")
    
    print("\nTriage complete!")


if __name__ == "__main__":
    main()
